src/c++/perf_analyzer/client_backend/openai/http_client.h

```diff
@@ -0,0 +1,176 @@
+  std::string cert;
+  // The format of the private key. By default it is KEY_PEM. See here for more
+  // details: https://curl.se/libcurl/c/CURLOPT_SSLKEYTYPE.html.
+  KEYTYPE key_type;
+  // The private key. See here for more details:
+  // https://curl.se/libcurl/c/CURLOPT_SSLKEY.html.
+  std::string key;
+};
+
+// HttpRequest object representing the context of a HTTP transaction. Currently
+// it is also designed to be the placeholder for response data, but how the
+// response is stored can be revisited later.
+// 'completion_callback' doesn't transfer ownership of HttpRequest, caller must
+// not keep the reference and access HttpRequest object after
+// 'completion_callback' returns
+class HttpRequest {
+ public:
+  HttpRequest(
+      std::function<void(HttpRequest*)>&& completion_callback,
+      const bool verbose = false);
+  virtual ~HttpRequest();
+
+  // Adds the input data to be delivered to the server, note that the HTTP
+  // request does not own the buffer.
+  void AddInput(uint8_t* buf, size_t byte_size);
+
+  // Helper function for CURL
+  // Copy into 'buf' up to 'size' bytes of input data. Return the
+  // actual amount copied in 'input_bytes'.
+  void GetNextInput(uint8_t* buf, size_t size, size_t* input_bytes);
+
+  // Buffer that accumulates the response body.
+  std::string response_buffer_;
+
+  size_t total_input_byte_size_{0};
+
+  // HTTP response code for the inference request
+  uint32_t http_code_{200};
+
+  std::function<void(HttpRequest*)> completion_callback_{nullptr};
+
+  // Pointer to the list of the HTTP request header, keep it such that it will
+  // be valid during the transfer and can be freed once transfer is completed.
+  struct curl_slist* header_list_{nullptr};
+
+ protected:
+  const bool verbose_{false};
+
+  // Pointers to the input data.
+  std::deque<std::pair<uint8_t*, size_t>> data_buffers_;
+};
+
+// Base class for common HTTP functionalities
+class HttpClient {
+ public:
+  enum class CompressionType { NONE, DEFLATE, GZIP };
+
+  virtual ~HttpClient();
+
+ protected:
+  void SetSSLCurlOptions(CURL* curl_handle);
+
+  HttpClient(
+      const std::string& server_url, bool verbose = false,
+      const HttpSslOptions& ssl_options = HttpSslOptions());
+
+  // Note that this function does not block
+  void Send(CURL* handle, std::unique_ptr<HttpRequest>&& request);
+
+ protected:
+  void AsyncTransfer();
+
+  bool exiting_{false};
+
+  std::thread worker_;
+  std::mutex mutex_;
+  std::condition_variable cv_;
+
+  // The server url
+  const std::string url_;
+  // The options for authorizing and authenticating SSL/TLS connections
+  HttpSslOptions ssl_options_;
+
+  using AsyncReqMap = std::map<uintptr_t, std::unique_ptr<HttpRequest>>;
+  // curl multi handle for processing asynchronous requests
+  void* multi_handle_;
+  // map to record ongoing asynchronous requests with pointer to easy handle
+  // or tag id as key
+  AsyncReqMap ongoing_async_requests_;
+
+  bool verbose_;
```

Comment:
nit: this isn't used

Reply:
I think that's ok. Good to have it there if needed for future code or to keep some breadcrumbs